[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33650 [not4land] debugging
* #33533 [not4land] Temporary workaround to get quantized ops
* #34572 [quant][graphmode] quantization support for aten::add
* #34518 [quant][graphmode] quantization work for prim::If
* #34571 [quant][graphmode][fix] use observed_values_ to check values are observed
* #34414 [quant][graphmode][refactor] insertObservers for Block
* #34411 [quant][graphmode][fix] Insert dequantize before use node
* #34349 [quant][graphmode][fix] preserve the type of original value when inserting dequant node
* #34348 [quant] Add dequantize.tensors
* #34347 [quant][graphmode] Add quantization support for aten::dropout
* #34346 [quant][graphmode] Add quantization support for aten::cat
* #34345 [quant][graphmode] Add prim::ListConstruct to general op handling
* #34315 [quant][graphmode] insert quant/dequant work for duplicated debugName
* #34314 [quant][graphmode][refactor] Change QParamMap to QParamVector
* #33854 [quant][graphmode] Quantization pattern for aten::linear
* #33532 [quant][graphmode] Add quantization pattern for quantized::add_relu
* #33279 [quant][graphmode] Add quantized conv2d-relu fusion pattern
* **#33927 [quant][graphmode] Add Finalize function that inlines graph and produce quantized ops**

Summary:

Test Plan:
test will be added in later PRs

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20354879](https://our.internmc.facebook.com/intern/diff/D20354879)

jerryzh168 committed:
[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops

Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
test will be added in later PRs

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
test will be added in later PRs

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
test will be added in later PRs

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
test will be added in later PRs

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
test will be added in later PRs

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
test will be added in later PRs

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/_quantize_script.py:
```diff
@@ -81,8 +81,7 @@ def convert_script(model, inplace=False):
     model = wrap_cpp_module(torch._C._jit_pass_insert_quant_dequant(model._c, 'forward', False))
-    if 'fbgemm' in torch.backends.quantized.supported_engines:
-        torch._C._jit_pass_insert_prepack_unpack(model._c)
+    model = wrap_cpp_module(torch._C._jit_pass_quant_finalize(model._c))
```
Please add an option to convert_script to enable/disable finalize so that users have the ability to dump the model prior to finalization. To make it simple for users, suggest
```python
convert_script(model, inplace=False, debug=False):
 
  if not debug:
       model = wrap_cpp_module(torch._C._jit_pass_quant_finalize(model._c))
```

raghuramank100 replied:
It is important to add this option to both quantize and convert script so that users can examine a model with quant/dequant ops prior to freezing and inlining.

raghuramank100 commented on torch/quantization/_quantize_script.py:
```diff
@@ -81,8 +81,7 @@ def convert_script(model, inplace=False):
     model = wrap_cpp_module(torch._C._jit_pass_insert_quant_dequant(model._c, 'forward', False))
-    if 'fbgemm' in torch.backends.quantized.supported_engines:
-        torch._C._jit_pass_insert_prepack_unpack(model._c)
+    model = wrap_cpp_module(torch._C._jit_pass_quant_finalize(model._c))
```
Also, can you add tests for convert_script with and without debug=True to ensure that this API is tested?

jerryzh168 replied:
tests are added in later PR: https://github.com/pytorch/pytorch/pull/33895/files

raghuramank100 replied:
For completeness, would it be possible to move them into this PR?

raghuramank100 reviewed (changes requested):
Please see comments, thanks!

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
test will be added in later PRs

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
test will be added in later PRs

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20354879](https://our.internmc.facebook.com/intern/diff/D20354879)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
test will be added in later PRs

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20354879](https://our.internmc.facebook.com/intern/diff/D20354879)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
test will be added in later PRs

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20354879](https://our.internmc.facebook.com/intern/diff/D20354879)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add Finalize function that inlines graph and produce quantized ops"


Summary:

Test Plan:
test will be added in later PRs

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20354879](https://our.internmc.facebook.com/intern/diff/D20354879)

[ghstack-poisoned]
